### PR TITLE
Writing appi tests-restful-booker

### PR DESCRIPTION
**Pull request by AlexVoinitch**

### CODE REVIEW

File: cypress/fixtures/booking-data.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (IvanZdanovich) [inline cypress/fixtures/booking-data.json]: do no use fixture files for storing test-data. use a specific js file and import it directly in the test file

File: cypress/integration/api/booking.api.spec.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        expect(response.status).to.eq(200);
+        expect(response.body).to.have.property('bookingid').to.be.a('number');
+
+        newBookingId = response.body.bookingid;
+
+        expect(response.body.booking.firstname).to.eq(testData.validBookingData.firstname);
+      });
+    });
+
+    it('should return 200(BUG - should be 500) when creating booking with invalid dates (Negative)', function () {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: please think about naming conventions for test block title. Apply it consistently.

File: cypress/support/api/booking-data.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// cypress/support/api/booking-data.js
+
```
Comment (IvanZdanovich) [inline cypress/support/api/booking-data.js]: now i see the test data file. good. get rid of the fixture file in that case

File: cypress/support/api/booking-data.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+const defaultBooking = {
+  firstname: 'Alexei',
+  lastname: 'Voinitch',
+  totalprice: 111,
+  depositpaid: true,
```
Comment (IvanZdanovich) [inline cypress/support/api/booking-data.js]: please provide a proper structure for your project, store separatly commands, test-data, constants like endpoints etc

File: cypress/integration/api/booking.api.spec.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
+// cypress/integration/api/booking.api.spec.js
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js:1]: Please review and address all Copilot comments—they are valid.

**Review by IvanZdanovich — CHANGES_REQUESTED** [review]

File: cypress/integration/api/booking.api.spec.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+describe('Booking Endpoint: Full CRUD Cycle', function () {
+  let newBookingId;
+
+  before(function () {
+    cy.userManagement__getUserDataByRole(userRoles.ADMIN_API).then((userData)=>{
+      cy.api__getAuthToken(userData);
+    });
+  });
+
+  context('Booking POST: Create Booking (Positive & Negative)', function () {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: Please review and try to improve the naming of describe, context and it blocks to make them more readable and clear

File: cypress/integration/api/booking.api.spec.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        expect(response.status).to.eq(200);
+        expect(response.body).to.have.property('bookingid').to.be.a('number');
+
+        newBookingId = response.body.bookingid;
+
+        expect(response.body.booking.firstname).to.eq(bookingData.validBookingData.firstname);
+      });
+    });
+
+    it('should return 200 status code for invalid checkin/checkout dates (BUG)', function () {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: try to make api tests more descrive - do not join test with different conditions (positve and negative) under the one context block, make title more informative, verbose

File: cypress/integration/api/booking.api.spec.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  });
+
+  context('Booking DELETE: Delete Booking and Verify Deletion', function () {
+    it('should successfully delete the booking and return 201 status', function () {
+      cy.api__deleteBooking(newBookingId).then((response) => {
+        expect(response.status).to.eq(201);
+      });
+    });
+
+    it('should return 404 after successful deletion (Verification)', function () {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: It should be in the next context block with a condition about a non-existing item

File: cypress/support/api/auth.api.commands.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+      method: 'POST',
+      url: `${API_URLS.BASE_URL}${API_URLS.ENDPOINTS.AUTH}`,
+      body: credentials,
+      headers: {
+        'Content-Type': 'application/json',
+        Accept: 'application/json',
+      },
+    })
+    .then((response) => {
+      expect(response.status).to.eq(200);
```
Comment (IvanZdanovich) [inline cypress/support/api/auth.api.commands.js]: We do not need verification of code placed inside the command itself; verification logic should be placed in tests

File: cypress/support/api/booking.api.commands.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
+// cypress/support/api/booking.api.commands.js
+
+import { BASE_URL, ENDPOINTS } from '../constants/api-urls';
```
Comment (IvanZdanovich) [inline cypress/support/api/booking.api.commands.js]: These URLs are accessible via global variables

File: cypress/support/api/booking.api.commands.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    headers: {
+      'Content-Type': 'application/json',
+      Accept: 'application/json',
+    },
+    body: bookingBody,
+    ...options,
+  });
+});
+
+Cypress.Commands.add('api__getBookingById', (bookingId, options = {}) => {
```
Comment (IvanZdanovich) [inline cypress/support/api/booking.api.commands.js]: 'api__' prefix is not informative

File: cypress/support/api/booking.api.commands.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    url: `${BASE_URL}${ENDPOINTS.BOOKING}/${bookingId}`,
+    headers: {
+      Accept: 'application/json',
+    },
+    ...options, // for sending failOnStatusCode: false
+  });
+});
+
+Cypress.Commands.add('api__updateBooking', (bookingId, updateBody) => {
+  const token = Cypress.env('authToken');
```
Comment (IvanZdanovich) [inline cypress/support/api/booking.api.commands.js]: token should be passed as a parameter, because in most cases you will work with multiple roles

File: docs/naming-conventions.md [hunk 13]
```diff
@@ -26,7 +28,32 @@ When referencing components in Cypress selectors, use descriptive names that ref
...
+## API Testing Conventions
+
+### 1. File Naming
+
+* **Format:** `[endpointName].api.spec.js`
+* **Example:** `ping.api.spec.js`, `booking.api.spec.js`
+
+### 2. Custom Command Naming
+
+* **Format:** `api__[method][EntityName]`
```
Comment (IvanZdanovich) [inline docs/naming-conventions.md]: Try to improve naming conventions to make them more clear and informative

**Review by IvanZdanovich — CHANGES_REQUESTED** [review]

File: cypress/integration/api/booking.api.spec.js (hunk 8 above)
Comment (AlexVoinitch) [reply cypress/integration/api/booking.api.spec.js]: done

File: cypress/support/api/auth.api.commands.js (hunk 9 above)
Comment (AlexVoinitch) [reply cypress/support/api/auth.api.commands.js]: done

File: cypress/support/api/booking.api.commands.js (hunk 11 above)
Comment (AlexVoinitch) [reply cypress/support/api/booking.api.commands.js]: Changed to format "domain-driven naming"

File: cypress/support/api/booking.api.commands.js (hunk 12 above)
Comment (AlexVoinitch) [reply cypress/support/api/booking.api.commands.js]: done

File: cypress/support/api/ping.api.commands.js [hunk 14]
```diff
@@ -0,0 +1,11 @@
+// cypress/support/api/ping.api.commands.js
+
+import { BASE_URL, ENDPOINTS } from '../constants/api-urls';
+
+Cypress.Commands.add('health__getCheck', () => {
+  return cy.request({
+    method: 'GET',
+    url: `${BASE_URL}${ENDPOINTS.PING}`,
+    failOnStatusCode: true,
```
Comment (IvanZdanovich) [inline cypress/support/api/ping.api.commands.js]: The failOnStatusCode flag is set by default

File: cypress/support/api/booking.api.commands.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      'Content-Type': 'application/json',
+      Accept: 'application/json',
+      Cookie: `token=${token}`,
+    },
+    body: updateBody,
+  });
+});
+
+Cypress.Commands.add('booking__delete', (bookingId, token = Cypress.env('authToken')) => {
+  if (!token) throw new Error('Auth token is required for booking__delete');
```
Comment (IvanZdanovich) [inline cypress/support/api/booking.api.commands.js]: There’s no need for an explicit token check, because negative scenarios must also be validated, and the API under test should correctly handle invalid or missing tokens.

File: cypress/support/api/ping.api.commands.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
+// cypress/support/api/ping.api.commands.js
+
+import { BASE_URL, ENDPOINTS } from '../constants/api-urls';
```
Comment (IvanZdanovich) [inline cypress/support/api/ping.api.commands.js]: There’s no need to create a separate variable for the base URL, since it’s already defined in the Cypress config and can be accessed directly.

File: cypress/integration/api/booking.api.spec.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+  context('When user performs a partial update (PATCH)', function () {
+    it('Then it should return 200 OK and reflect the updated firstname', function () {
+      const updateBody = { firstname: bookingData.updatedBookingData.firstname };
+      cy.booking__update(newBookingId, updateBody).then((response) => {
+        expect(response.status).to.eq(200);
+        expect(response.body.firstname).to.eq(bookingData.updatedBookingData.firstname);
+      });
+    });
+  });
+
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js:64]: The context title ‘When user deletes booking’ doesn’t provide enough detail about the conditions being tested. Please use more descriptive and meaningful titles for both context blocks and their contained tests. Also, separate negative scenarios into their own dedicated context block to improve clarity and structure

File: cypress/integration/api/booking.api.spec.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    it('Then it should return 404 Not Found for a non-existent ID', function () {
+      cy.booking__getById(-1, { failOnStatusCode: false }).then((response) => {
+        expect(response.status).to.eq(404);
+      });
+    });
+  });
+
+  context('When user performs a partial update (PATCH)', function () {
+    it('Then it should return 200 OK and reflect the updated firstname', function () {
+      const updateBody = { firstname: bookingData.updatedBookingData.firstname };
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: I suggest using a prefix with a predefined structure. It would be helpful to include details such as the module, endpoint, and action — for example: Booking.Booking.RetrieveDetails.GET

It will help when the project will be extended into several modules.

File: cypress/integration/api/booking.api.spec.js [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        newBookingId = response.body.bookingid;
+        expect(response.body.booking.firstname).to.eq(bookingData.validBookingData.firstname);
+      });
+    });
+
+    it('Then the system should return 200 for invalid dates (Known BUG)', function () {
+      const invalidBody = {
+        ...bookingData.validBookingData,
+        bookingdates: { checkin: '2026-01-29', checkout: '2026-01-26' },
+      };
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: could you prepare test data instance in the test data?

File: cypress/integration/api/booking.api.spec.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
+// cypress/integration/api/booking.api.spec.js
+
+import { bookingData } from '../../support/test-data/booking-data';
+
+describe('Booking API: Given the Booking endpoint is available', function () {
+  let newBookingId;
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: This ID should be stored in the test data

File: cypress/integration/api/booking.api.spec.js [hunk 21]
```diff
@@ -3,68 +3,57 @@
...
   before(function () {
     return cy.userManagement__getUserDataByRole(userRoles.ADMIN_API).then((userData) => {
       cy.auth__getToken(userData);
     });
   });
 
-  context('When user attempts to create booking', function () {
-    it('Then the system should create a new record and return 200 OK (Positive)', function () {
+  context('Booking.Booking.PositiveScenarios', function () {
+    it('Booking.Booking.Create.POST: Then the system should create a new record and return 200 OK', function () {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: Please provide a separate context block for each it block. The context titles should clearly describe the conditions or parameters used when testing each specific case and endpoint

File: cypress/integration/api/booking.api.spec.js [hunk 22]
```diff
@@ -0,0 +1,61 @@
+// cypress/integration/api/booking.api.spec.js
+
+import { bookingData } from '../../support/test-data/booking-data';
+
+describe('Booking API: Given the Booking endpoint is available', function () {
+  before(function () {
+    return cy.userManagement__getUserDataByRole(userRoles.ADMIN_API).then((userData) => {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: why return keyword is used here?

File: cypress/integration/api/booking.api.spec.js [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+
+    it('Booking.Booking.Delete.DELETE: Then the system should return 201 Created for a successful deletion', function () {
+      cy.booking__delete(bookingData.dynamicData.newBookingId).then((response) => {
+        expect(response.status).to.eq(201);
+      });
+    });
+  });
+
+  context('Booking.Booking.NegativeScenarios', function () {
+    it('Booking.Booking.Create.POST: Then the system should return 200 even with invalid dates (Known BUG)', function () {
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: where is the link on the bug?

File: cypress/integration/api/booking.api.spec.js [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+        expect(response.status).to.eq(200);
+        bookingData.dynamicData.newBookingId = response.body.bookingid;
+        expect(response.body.booking.firstname).to.eq(bookingData.validBookingData.firstname);
+      });
+    });
+
+    it('Booking.Booking.GetDetails.GET: Then the system should return 200 OK for an existing ID', function () {
+      cy.booking__getById(bookingData.dynamicData.newBookingId).then((response) => {
+        expect(response.status).to.eq(200);
+        expect(response.body.firstname).to.eq(bookingData.validBookingData.firstname);
```
Comment (IvanZdanovich) [inline cypress/integration/api/booking.api.spec.js]: can we check all the fields from the response?

File: docs/naming-conventions.md [hunk 25]
```diff
@@ -26,7 +28,33 @@ When referencing components in Cypress selectors, use descriptive names that ref
...
+* **Examples:** `cy.booking__create()`, `cy.auth__getToken()`, `cy.booking__getById()`
+
+### 3. Test Titles (Gherkin Style)
+
+To make tests readable and descriptive, use the **Given / When / Then** approach:
+
+* **Describe (Given):** The initial context or precondition.
+  * *Example:* `describe('Booking API: Given the user has valid credentials', ...)`
+* **Context (When):** The specific action or scenario being tested. Keep positive and negative scenarios in separate context blocks.
+  * *Example:* `context('When the user attempts to create a booking with valid data', ...)`
```
Comment (IvanZdanovich) [inline docs/naming-conventions.md:55]: define the naming convention for test blocks and describe here.

File: cypress/support/test-data/booking-data.js [hunk 26]
```diff
@@ -0,0 +1,32 @@
+// cypress/support/test-data/booking-data.js
+
+export const bookingData = {
+  dynamicData: {
+    newBookingId: null,
+  },
+  validBookingData: {
+    firstname: 'Alexei',
+    lastname: 'Voinitch',
+    totalprice: 111,
```
Comment (IvanZdanovich) [inline cypress/support/test-data/booking-data.js]: could we add some randomisation for test data?

**Review by IvanZdanovich — CHANGES_REQUESTED** [review]

File: cypress/support/test-data/booking-data.js [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+const randomId = () => Math.floor(Math.random() * 1000);
+
+export const bookingData = {
+  dynamicData: {
+    newBookingId: null,
+  },
+
+  validBookingData: {
+    firstname: `Alexei${randomId()}`,
+    lastname: `Voinitch${randomId()}`,
```
Comment (ivan-zdanovich-citation) [inline cypress/support/test-data/booking-data.js]: could be randomized using all allowed characters not numbers from 0 to 1000

File: cypress/support/test-data/booking-data.js [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+    totalprice: 100 + randomId(),
```
Comment (ivan-zdanovich-citation) [inline cypress/support/test-data/booking-data.js]: why not to randomize numbers from 1 to 100000?

File: cypress/support/test-data/booking-data.js [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+    depositpaid: true,
+    bookingdates: {
+      checkin: '2026-01-26',
+      checkout: '2026-01-29',
+    },
+    additionalneeds: 'Breakfast',
+  },
+
+  invalidDatesBooking: {
+    firstname: 'Bug',
```
Comment (ivan-zdanovich-citation) [inline cypress/support/test-data/booking-data.js]: randomize again and again. more and more fields it will help to avoid test efficiency degradation, explore or confirm the limitations and requirements of the app

**Review by ivan-zdanovich-citation — APPROVED** [review]

**Review by IvanZdanovich — DISMISSED** [review]

File: cypress/support/test-data/booking-data.js (hunk 27 above)
Comment (AlexVoinitch) [reply cypress/support/test-data/booking-data.js]: Done. Added random string generator and expanded number range to 100,000.

**Review by IvanZdanovich — APPROVED** [review]
